### ISSUE TITLE
When display is set to block, field height grows unexpectedly

### ISSUE THREAD

**Issue by web-padawan** [vaadin/vaadin-text-field#201]
The default `display` value for element is `inline-flex`, however, it seems quite common solution to set  it to `block` in order to expand for 100% width. But once we do it, the `::before` pseudo element gets misaligned causing the following result: 

<img width="1104" alt="screen shot 2018-02-23 at 19 18 42" src="https://user-images.githubusercontent.com/10589913/36607368-1d985532-18cf-11e8-8c13-b18ed17af524.png">

**Comment by jouni** [comment]
Not sure how to mitigate this issue. Obviously, you should use `display: flex;` instead of `display: block;`, but it seems impossible to communicate that effectively to users.

**Comment by web-padawan** [comment]
Users should be able to set display to block as well, IMO. This is even worse than using `position: relative !important` in button :(

**Comment by jouni** [comment]
Yeah, very true. I can try to prototype if there’s a way to fix the vertical alignment issues without flexbox.

**Comment by jouni** [comment]
Also, the button styles should be prototyped more as well.

**Comment by jouni** [comment]
Prototyped this quickly: https://github.com/vaadin/vaadin-text-field/commit/0507dbb1d36a54812c44d8872f8e763a502f439f

I’m not too fond of this, though, mainly because it breaks how text-area supports `height` set on the host element. The `line-height: 0;` on the host is also ugly, but that was the only way I could quickly think of to make the first line-box inside the host to collapse smaller than the label element height.

**Comment by jouni** [comment]
About the button (although this is probably the wrong place for this discussion), as it’s also about default layout properties: I think we should drop the `!important` from there `position: relative;` on the host. The only fix I can think of is to move the native `<button>` element inside `.vaadin-button-container` and make the container relative instead.

Note, that Lumo still expects the host to always be positioned, as it uses pseudo elements on the host for the hover and activation effects.

**Comment by web-padawan** [comment]
I'm wondering if the pseudo elements on the host element should be considered as not to be changed by user. It is rather unexpected to change these, although I can imagine the attempt to do some customization. If we follow this convention, we probably should document it. This makes sense especially because of no pseudo elements allowed on the native `input` element.

About the text area, and the alignment concept in general: this looks like a feature, but personally I would rather expect to rely on native CSS features for alignment, either using `display: inline-block` and `vertical-align`, or `display: flex` on the container. The element trying to align itself currently breaks at least first assumption.

**Comment by jouni** [comment]
> I'm wondering if the pseudo elements on the host element should be considered as not to be changed by user.  It is rather unexpected to change these, although I can imagine the attempt to do some customization. If we follow this convention, we probably should document it. This makes sense especially because of no pseudo elements allowed on the native input element.

Yes, we should have better documentation what types of customizations are supported on the elements. I have also a todo hanging that we should document that we do not guarantee the stability of the exact CSS selectors and properties we have in the themes – instead, we aim to document (in release notes) any changes that can potentially break customizations.

> About the text area, and the alignment concept in general: this looks like a feature, but personally I would rather expect to rely on native CSS features for alignment, either using display: inline-block and vertical-align, or display: flex on the container. The element trying to align itself currently breaks at least first assumption.

Text area? I assume you mean `vaadin-text-field` in general.

Adjusting `vertical-alignment` is still possible on the host, even if we align it by default to the text-input text.

I’m currently leaning on not doing any fixes at the moment, and instead figuring out where should we start documenting these kinds of styling concerns.

**Comment by jouni** [comment]
One more workaround/fix to consider:

```css
:host {
  white-space: nowrap;
}

.vaadin-text-field-container {
  display: inline-flex;
  vertical-align: top;
}
```

This allows any display value for the host.

**Comment by web-padawan** [comment]
@jouni this looks better but seems to still require keeping `:host::before`, right? And when setting `display: block`, that pseudo element persists and makes field misaligned from the left side.

**Comment by jouni** [comment]
Yes, the alignment pseudo-element is still required. We could move that to an actual element in the template – would probably make sense.

And yes, when setting `display: block` on the host, the alignment element and the internal `.vaadin-text-field-container` element are still on the same line and the alignment element sets the baseline. Though, I only tested this quickly in Chrome, so not absolutely certain.
